presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftIndexPageSource.java

```diff
@@ -0,0 +1,392 @@
+            splits.addAll(batch.getSplits());
+            // check if it's possible to request more splits
+            if (batch.getNextToken() != null) {
+                // can get more splits, send request
+                splitFuture = sendSplitRequest(splitsClient, batch.getNextToken());
+                statusFuture = toCompletableFuture(nonCancellationPropagating(splitFuture));
+                return null;
+            }
+            else {
+                // no more splits
+                splitFuture = null;
+                statusFuture = null;
+                haveSplits = true;
+                splitsClient.close();
+                splitsClient = null;
+            }
+        }
+        checkState(haveSplits, "must have splits at this point");
+
+        // check if any data requests were started
+        if (dataSignalFuture == null) {
+            // no data requests were started, start a number of initial requests
+            checkState(contexts.isEmpty() && dataRequests.isEmpty(), "must have no running splits at this point");
+            if (splits.isEmpty()) {
+                // all done: no splits
+                finished = true;
+                return null;
+            }
+            for (int i = 0; i < min(lookupRequestsConcurrency, splits.size()); i++) {
+                PrestoThriftSplit split = splits.get(splitIndex);
+                splitIndex++;
+                RunningSplitContext context = new RunningSplitContext(openClient(split), split);
+                sendDataRequest(context, null);
+            }
+            dataSignalFuture = whenAnyComplete(dataRequests);
+            statusFuture = toCompletableFuture(nonCancellationPropagating(dataSignalFuture));
+        }
+        // check if any data request is finished
+        if (!dataSignalFuture.isDone()) {
+            // not finished yet
+            return null;
+        }
+
+        // at least one of data requests completed
+        ListenableFuture<PrestoThriftPageResult> resultFuture = getAndRemoveNextCompletedRequest();
+        RunningSplitContext resultContext = contexts.remove(resultFuture);
+        checkState(resultContext != null, "no associated context for the request");
+        PrestoThriftPageResult pageResult = getFutureValue(resultFuture);
+        Page page = pageResult.toPage(outputColumnTypes);
+        if (page != null) {
+            completedBytes += page.getSizeInBytes();
+        }
+        if (pageResult.getNextToken() != null) {
+            // can get more data
+            sendDataRequest(resultContext, pageResult.getNextToken());
+            dataSignalFuture = whenAnyComplete(dataRequests);
+            statusFuture = toCompletableFuture(nonCancellationPropagating(dataSignalFuture));
+            return page;
+        }
+
+        // split finished, closing the client
+        resultContext.close();
+
+        // are there more splits available
+        if (splitIndex < splits.size()) {
```

Comment:
It will create unnecessary wrapping futures when doing initial setup. However, I extracted inner parts of the method, so it should be easier to read now.